### ISSUE TITLE
CSS reformat problem

### ISSUE THREAD

**Issue by BeckyDTP** [Sigil-Ebook/Sigil#328]
Hello,

I know you are preparing version 0.9.9, but maybe you can take a look at it.

One line CSS file with media query.
Reformat CSS and a strange result.

[Here you can clearly see it.](https://i.imgur.com/ND1Gp4a.png)

**Comment by dougmassay** [comment]
I see it. Weird. It's not a feature I use much, so I've never noticed. We'll look into it. Unfortunately, 0.9.9 cannot be held up to accommodate it. Any fix for this will be in a later release.

I _will_ see if a plugin to reformat css (including media queries) can be slapped together as a workaround in the mean-time.

Thanks.

**Comment by kevinhendricks** [comment]
Yes we will look into this bug for the next release after 0.9.9.
Thank you for your bug report.

Kevin

**Comment by kevinhendricks** [comment]
Sigil/src/Misc/CSSInfo.cpp does not appear to grok media queries at all.  We really should replace this CSS parser class with calls to a real CSS parser library built either in python, C, or C++ as CSS has moved on since this parser was designed and written.

We do have a CSS parser in the plugin library, so using python to fix this with python3lib calls  probably makes themost sense unless we can find a good opensource CSS parser lib to pull into Sigil.

**Comment by wrCisco** [comment]
I just stumbled upon [MyCSS](https://github.com/lexborisov/mycss), C css parser of the Modest engine, which looks promising (there is also a [python binding](https://github.com/rushter/selectolax) of the engine). Remembering issue #219 I could try to do some testing and patch it into my version of Sigil, but I can't guarantee about timing...

**Comment by kevinhendricks** [comment]
Sounds good to me.  I took a peek at it and I think that library relies on other my* component libraries so it is not quite standalone.  We could theoretically extract and combine the parts we need into one css library if need be.

**Comment by kevinhendricks** [comment]
Whoops hit the close by mistake.

**Comment by wrCisco** [comment]
Yes, the up to date version of mycss is inside the [Modest repository](https://github.com/lexborisov/Modest) and seems to be dependent upon myhtml, myencoding and mycore. For now I built the entire engine and I'm doing some experiment just to wrap my head around its API...

**Comment by kevinhendricks** [comment]
How did this work out?  I am thinkng if it looks good at all that adding this css parser to Sigil is orobably a good way to go.

**Comment by kevinhendricks** [comment]
This bug has been reported again privately via pm on Mobileread and really needs to get fixed for the next release.  We can not have Prettying and Reformating CSS making a mess of things due to media queries.

The problem is that Sigil's current CSS parser was written long before media queries were thought of.  The Sigil css  parsing code itself is just too simplistic.  I have examined Modest (mycss) and a few other C++ or C based CSS parsers (libcss, etc) but most of these have too many other outside requirements that would make adding them to Sigil a bit of an issue.  Modest is small and builds out of the box on Linux and OSX (and hopefully Windows) but there is very little documentation on using it or the interface itself.  It would also mean adding in yet another buffer, string allocation set of C code to deal with, have buffer overruns with, make mistakes with, etc.

@wrCisco,  Since your (wrCisco)'s  unused css removal plugin has found and worked around a few bugs in python's cssutils package and pretty much does what we need.  It just seems logical to use python/cssutils as a solution for Sigil's outdated css parsing code permanently (or until a C++ or C library that is simple and stands alone can be found).

So before the next release of Sigil we will look into adapting wrCiscos's cssutils plugin code to work internally in Sigil.

This is just a head's up in how we plan to address this and related issues unless someone has a better idea.

If so, please let me know.

Thanks

**Comment by kevinhendricks** [comment]
Based on some recently provided test cases, the latest version of cssutils will not properly parse @media rules that use device-aspect-ratio or aspect-ratio as its tokenizer barfs on the "/" that is used to identify the ratio.  Also cssutils does not recognize @supports as a @ keyword.

To make it orse, there is a bug report about aspect-ratios filed by Kovid almost 2 years ago with nothing being done to date.

So we may need to explore other options for a css parser or pull in our own copy of cssutils like we do with bs4.

Ideas and thoughts welcome on the best way forward.

**Comment by wrCisco** [comment]
Hi, sorry for the late reply.
Months ago, I did some testing with mycss, but it didn't seem production ready, at least in regard to the serializer.
If you want to work with my plugin, you're welcome (the version here on github is slightly more updated than the one on mobileread), but as you noted, there are some unresolved issues with cssutils.
Did you see [tinycss2](https://github.com/Kozea/tinycss2)? I'm not sure that's right for Sigil's needs, but seems a lot simpler to patch than cssutils if needed.
If you need help with testing and/or patching, I'll try not to disappear as in the last few months...

**Comment by kevinhendricks** [comment]
Okay, here is a very very tentative patch for supporting aspect-ratio values in media queries in cssutils:

First I grabbed the latest cssutils-1.0.2.tar.gz and unpacked it.  Then ran "2to3" on both the cssutils and encutils, to get it to function under python3.

Then I made the following changes:
```
--- cssproductions.py.keep	2018-12-14 13:10:32.000000000 -0500
+++ cssproductions.py	2018-12-14 11:58:40.000000000 -0500
@@ -74,6 +74,7 @@
 
     ('S', r'{s}+'), # 1st in list of general productions
     ('URI', r'{U}{R}{L}\({w}({string}|{url}*){w}\)'),
+    ('RATIO', r'{w}[0-9]+{w}\/{w}[0-9]+'),
     ('UNICODE-RANGE', r'{U}\+[0-9A-Fa-f?]{1,6}(\-[0-9A-Fa-f]{1,6})?'),
     ('IDENT', r'{ident}'),
     ('FUNCTION', r'{ident}\('),
--- prodparser.py.keep	2018-12-14 13:15:40.000000000 -0500
+++ prodparser.py	2018-12-14 13:16:26.000000000 -0500
@@ -801,6 +801,16 @@
                     )
 
     @staticmethod
+    def ratio(stop=False, nextSor=False):
+        "positive integer / positive integer"
+        return Prod(name='ratio',
+                    match=lambda t, v: t == PreDef.types.RATIO,
+                    toSeq=lambda t, tokens: (t[0], t[1].lower()),
+                    stop=stop,
+                    nextSor=nextSor
+                    )
+
+    @staticmethod
     def variable(toSeq=None, stop=False, nextSor=False, toStore=None):
         return Prod(name='variable',
                     match=lambda t, v: 'var(' == cssutils.helper.normalize(v),
--- css/value.py.keep	2018-12-14 13:00:31.000000000 -0500
+++ css/value.py	2018-12-14 13:03:57.000000000 -0500
@@ -107,7 +107,7 @@
                 [ NUMBER S* | PERCENTAGE S* | LENGTH S* | EMS S* | EXS S* |
                   ANGLE S* | TIME S* | FREQ S* ]
               | STRING S* | IDENT S* | URI S* | hexcolor | function
-              | UNICODE-RANGE S*
+              | RATIO S* | UNICODE-RANGE S*
               ;
             function
               : FUNCTION S* expr ')' S*
@@ -264,6 +264,7 @@
         prods = Choice(PreDef.hexcolor(stop=True),
                        PreDef.ident(stop=True),
                        PreDef.string(stop=True),
+                       PreDef.ratio(stop=True),
                        PreDef.unicode_range(stop=True),
                        )
         ok, seq, store, unused = ProdParser().parse(cssText, 'Value', prods)
```

which defined a new value type called ratio in cssproductions, added it to the PreDef class in value.py so it could be identified, and added it to prodparser.py

When I run the simple testcase provided by Kovid's bug report:

python3 -c "import cssutils; cssutils.parseString('@media (aspect-ratio:1280/800) { body { color: red }}')"
 
I now get no errors, instead of:

ERROR   MediaQuery: Missing token for production expression END: ('CHAR', u'/', 1, 26)
ERROR   MediaList: No match: ('NUMBER', u'800', 1, 27)
ERROR   MediaQuery: No content.

This extends the Value class to recognize and allow ratios.  Perhaps I should have kept the old Value class and instead created a new extended value class that would only be allowed in media queries.  Who knows?

I have no idea if ratios can be used in any other place.  The basic css parse/syntax docs do not really specify things in media queries well at all or really any at rule well at all.  That is why tinycss2 went to the simpler approach but ...

Please give these changes a try in cssutils and if they work, I will work on the @supports rule next.

Then we can make all of our changes in our own copy of cssutils and absorb it into our tree or move to Kovid's tinycss version.

**Comment by dougmassay** [comment]
Would the plan be to move cssutils from the bundled python environment to the same place our patched bs4 module resides? If so, we'll have to be careful if we change the name like we did with sigil_bs4. There's some plugins that use the bundled cssutils. A name change would require them to update their plugins. Not that big of a deal, but if it could be avoided...

/plugin_launcher/python (the home of sigil_bs4) is already prepended to the existing sys.path, so there shouldn't be any confusion with installed system versions of cssutils if we don't change the name (thinking of Linux and externally configured pythons primarily.

Or were you thinking using the patched version strictly for Sigil's own embedded python routines and leaving the standard cssutils in the plugin environment?

**Comment by kevinhendricks** [comment]
We could go either way.  One solution might be to include it but keep the exact name and remove it from the bundled python list.  That way our version will be found first I think (all it will have is bug fixes over the older official version).    I do not plan on making any changes to the official interface methods, just fix bugs that do not seem to get fixed upstream.

**Comment by dougmassay** [comment]
That sounds like the best plan to me (keep the same name and remove it from the bundled python).

I'll see if I can dig out some meda query-laden css and test how cssutils will parse it. I certainly won't be the definitive voice of whether or not it's all working "correctly," though. Complex css really isn't my strong suit. ;)

**Comment by wrCisco** [comment]
Are the four you posted the only changes you applied to cssutils? I tested it (both on the cssutils wheel already for python3 and the tarball) and I keep getting the same error as Kovid on the testcase.

Another place where there can be ratios in CSS syntax is inside the function calc (but that has its own production functions in cssutils). AFAIK the aspect-ratio is only used in media queries.

**Comment by kevinhendricks** [comment]
I probably missed something.  I will check.

**Comment by kevinhendricks** [comment]
Checked and I did not miss any.  These changes are enough to fix the issue with Kovid's testcase.  Only those 3 files were changed.  My guess is you are finding another cssutils somewhere in your python path.  I am testing with a fresh python3 installation with no cssutils installed except for the version local to my testcase.

Here is an even simpler patch created by a recursive diff that should work too that restricts the ratio to being inside media queries.

[changes2.patch.zip](https://github.com/Sigil-Ebook/Sigil/files/2682239/changes2.patch.zip)

**Comment by kevinhendricks** [comment]
Here is my exact work sequence with a python3 interpreter with no cssutils installed at all:

tar -zxvf cssutils-1.0.2.tar.gz
unzip changes2.patch.zip
2to3 --output-dir=cssutils-v3/cssutils -W -n cssutils-1.0.2/src/cssutils
2to3 --output-dir=cssutils-v3/encutils -W -n cssutils-1.0.2/src/encutils
cd cssutils-v3
patch -p0 < ../changes2.patch
python3 -c "import cssutils; cssutils.parseString('@media (aspect-ratio:1280/800) { body { color: red }}')"

**Comment by kovidgoyal** [comment]
As for sigil's plugin cssutils issues, that can be resolved by adding an importhook to the python environment to map cssutils imports to the new namespace.

**Comment by dougmassay** [comment]
> As for sigil's plugin cssutils issues, that can be resolved by adding an importhook to the python environment to map cssutils imports to the new namespace.

You're right, of course. Actually, from within our launcher.py we can do something as simple as:

`import whatever_we_call_the_new_cssutils`
`sys.modules['cssutils'] = whatever_we_call_the_new_cssutils`

The same could be done for our patched sigil_bs4 so that plugins could import using either the actual name, or the new namespaced one with the same results.

Thanks! I'll let you guys get back to cssutils forking/patching.

**Comment by kevinhendricks** [comment]
I am happy with the idea of forking cssutils and having a joint repo to maintain it.  Also the "python-css" name sounds fine to me as long as it does not already exist in pypi or a google search.  I could also see a "cssutils2" name if anyone were interested.  I will probably post my patch on the official cssutils site bug report to be nice but I have little hope for any kind of response given he ignored your earlier serialize.py fix for years now.

I see work being done in calibre towards python3 compatibility.  If so, we maintain our own fork of bs4 in python3 because of bugs and special uses, just in case you want to add that as jointly maintained as well.

**Comment by kovidgoyal** [comment]
I replaced beautifulsoup with the much faster, more accurate, HTML 5
parser for python I wrote a while ago --
https://github.com/kovidgoyal/html5-parser (this is based on gumbo and
actually includes many of the fixes you have in your gumbo fork).  

BeautifulSoup still exists in calibre, mainly for API compatibility
reasons, actual HTML parsing is done by html5-parser, which builds a BS
tree. So I think I'll pass on bs4, thanks.

Let's agree on a name and we can move forward. Candidates are:

css-parser, python-css, cssutils2, css_utils

this is in my preference ordering for the names. As far as I can tell,
none of them is taken on pypi.

**Comment by kevinhendricks** [comment]
Wow!  I had no idea you had forked gumbo as well.  Your version of gumbo has fixes and speed-ups Sigil will need to pull in based on just a quick glance.  As for bs4, since it is just a soup-like wrapper around another parser (ie a treebuilder like interface)  for html5lib, lxml, or gumbo, we fixed a few things to make our bs4 serialize a bit better, and provide it for our plugin interface so that people who are used to soup can use a parser more easily in their plugins.

Since official gumbo support seems to have been abandoned, would you consider adding your version of gumbo to the proposed shared repo?  I can then sync our version too it and help maintain that part.

As for a name for our cssutils fork, your "css-parser" choice would go well with your existing "html5-parser" name, so that works for me.

**Comment by kovidgoyal** [comment]
https://github.com/orgs/ebook-utils

is the new organization

**Comment by kovidgoyal** [comment]
feel free to upload the fork of cssutils with your patch

**Comment by kevinhendricks** [comment]
@wrCisco 
Glad to hear it now works!  And I am happy to have you on board to help. 
Thanks

**Comment by kovidgoyal** [comment]
I hate 2to3 as well. I much prefer polyglot code -- i.e. code that runs
on both. The approach I usually use is to run 2to3 in report mode and
then just change everything it points out to work on both via simple
polyglot module, such as polyglot.py in the calibre source code.

**Comment by kevinhendricks** [comment]
Yes your polyglot code is very much like our compatibility_utils.py code inside Sigil and KindleUnpack.  It is amazing how much time and duplication of effort has occurred to fix the same problems to make one code base that works for both.  I will take a shot at getting cssutils into some "polyglot" shape so that it runs on python2.7 and python3.x and then make the initial commit.

**Comment by kovidgoyal** [comment]
Great, let me know if you need a hand.

**Comment by kevinhendricks** [comment]
@kovidgoyal 
Do we have a minimum python 2.x version we need to support?  ie. 2.6 or 2.7 or earlier?  The reason I ask is the use of next(tokens) vs tokens.next() was backported to python2.6 but no earlier, so I need to special case it in the code if we want to support anything earlier than python 2.6 but nothing would be needed if we use next(tokens) for python 2.6, 2.7, and 3.x

**Comment by kovidgoyal** [comment]
I am fine with >= 2.7.9

**Comment by kevinhendricks** [comment]
This will still take a while since I had to fix a few bugs to get all of the unittests to pass under python3.x.  Having trouble getting the python2.x to do the same due to changes in code used in tests/basetest.py due to underlying differences in how os.path.abspath determines its current working directory from the python module that tries to load a file.

Perhaps, I will just plunge ahead tomorrow and take my best shot at making all of the test cases polyglot and then post what I have so others can see if they can figure out what is up.

**Comment by kovidgoyal** [comment]
Yeah go ahead and upload it, I will take a look this weekend.

**Comment by kevinhendricks** [comment]
Okay I pushed it.  The repo is:

https://github.com/ebook-utils/css-parser

The code should now be in dual PY2, PY3 mode.  I have not converted the test cases yet as I wanted to be able to verify correct behaviour with python 2.7 and python 3.x before merging the test code.  So inside src/cssutils you will find a tests folder (py3) and a tests2 folder (py2).

To run the tests, simply copy the dotestsPY*.sh to the src directory after checkout and run it or any specific test form it.

Using Python3.X, it now passes all tests except for test_util.py where the test itself is flawed as it looks for http:/example.com/test.css but that file does NOT exist.

Using Python2.x, it fails the same test_util.py test for the same reason.  It also fails one other test which is test_cssutilsimport.py as I get a TypeError: Item in "from list" not a string which is caused by this line (line 4) of that test file:  from cssutils import *
I think this is a sys.path issue but I am not sure ....  It passes all other tests with python2.7.

Instead of using a polyglot module, I made what few changes were needed local to each file to make each self-contained and easy to understand when viewed alone or playing round with in isolation.

I have already included an improved version of the changes2.patch to fix the aspect-ratio issue.
I have *NOT* added any other fixes we might have (the serialize fix and extending the media query types for Amazon, etc...

I have invited Sigil's co-maintainer @dougmassay to be an admin, and made sure @kovidgoyal was there as well.

@wrCisco if you want to work on this, just let me know and I will add you to the team with write privs.  

@BeckyDTP  please let me know your level of interest in working on this project as well.

@kovidgoyal what would be the best way to make your version of sigilgumbo/gumbo part of ebook-utils?

**Comment by kovidgoyal** [comment]
On Thu, Dec 20, 2018 at 08:46:47AM -0800, Kevin Hendricks wrote:
> Using Python2.x, it fails the same test_util.py test for the same reason.  It also fails one other test which is test_cssutilsimport.py as I get a TypeError: Item in "from list" not a string which is caused by this line (line 4) of that test file:  from cssutils import *
> I think this is a sys.path issue but I am not sure ....  It passes all other tests with python2.7.

I'll take care of fixing the tests. 

> 
> Instead of using a polyglot module, I made what few changes were needed local to each file to make each self-contained and easy to understand when viewed alone or playing round with in isolation.
> 
> I have already included an improved version of the changes2.patch to fix the aspect-ratio issue.
> I have *NOT* added any other fixes we might have (the serialize fix and extending the media query types for Amazon, etc...

We can do that after I have fixed the tests.

> @kovidgoyal what would be the best way to make your version of sigilgumbo/gumbo part of ebook-utils? 

Not sure, the obvious way is to either move sigil-gumbo/gumbo to
ebook-utils/gumbo or for me to create ebook-utils/gumbo 
with a copy of the gumbo code from html5-parser.

The problem is that html5-parser makes use of the fact that it is
python based to automatically run an extensive html5 test suite
including all the tests from html5lib. With pure gumbo, running those
tests is not as easy/convenient.

**Comment by kevinhendricks** [comment]
okay, lets wait on gumbo.  I think I grabbed all the bug fixes you made and can push them to Sigil.  That just leaves your speedups but they can come later.

**Comment by kevinhendricks** [comment]
@kovidgoyal As for the one test failure in python2.X:

TypeError: Item in "from list" not a string

You may already know this but it seems that python2 will only accept str classes (bytes) not unicode for the __all__ list of classes to import.  Obviously PY3 only accepts str (unicode).

So when we use from future import unicode_literals
and then it is followed by an __all__ being defined, and use of: 
   from cssutils import *
it will of course fail.

This is used in test_cssutilsimport.py and possibly by other users at some point.

To fix this we could remove unicode_literals from future and redo the complete polyglot process and deal with u' all throughout the code, or convert the __all__ list properly for each platform after it is defined.  

Something along the lines of the following would work. 

__all__ = ['css', 'stylesheets', 'CSSParser', 'CSSSerializer']
__all__ = [str(n) for n in __all__]

 I tried this approach on cssutils/__init__.py and now test_cssutilsimport.py now passes on python 2.x.

If we decide to go this approach, we would have to fix them all, which is easy enough to do.

That said, running the pep8 cleanup code, actually broke both PY2 and PY3 test_cssutilsimport.py tests because it moved all of the imports to the top of the file blindly.  I committed fixes to put them back where they belong.

**Comment by kovidgoyal** [comment]
Oh also, I will be working extensively on this codebase for the next couple of days, so things are likely to break temporarily. I intend to move the tests out of the cssutils package, and setup CI on Travis and appveyor for testing instead. I am also going to rename the cssutils package to css_parser and probably move encutils into a subpackage.

**Comment by kevinhendricks** [comment]
Sigil moved a while back to Python 3 only for plugins, so letting it break is fine with me.  It is easy to add back if it turns out to be needed later.

Thanks for letting me know about the breakage for the next week or so.   We are in no rush.  Please let us know when you think it is stable enough to start adding in things.

Thanks

**Comment by kovidgoyal** [comment]
You can now run the tests by using

python2 run_tests.py
python3 run_tests.py

all tests pass.

**Comment by kovidgoyal** [comment]
I'm done working on it, all the previously discussed items have been implemented. Feel free to add new fixes as you see fit and when you are done, we can discuss releasing it on pypi.

**Comment by kevinhendricks** [comment]
Wow, that's fast!  So are we going with css-parser or css_parser and if the latter should we change the project name too?

As for gumbo, I think we will stay separate as you may well want to have gumbo build lxml style nodes directly, whereas much of our code uses gumbo nodes directly in our C++ and only build lxml style trees for use in plugins.

That said, I will sync in most of your gumbo changes to date into our sigil-gumbo and add in a nesting level limit (which will require an addition to the options structure).

**Comment by kovidgoyal** [comment]
I prefer keeping the project name as css-parser and the python import name as css_parser (since you cannot use hyphens in import names). This is how html5-parser works as well (import name of html5_parser). 

Yeah I might someday have gumbo build lxml nodes directly for an even greater performance improvement.

**Comment by kevinhendricks** [comment]
@wrCisco 
The invite should be out there now.  Thanks for helping out!

**Comment by kevinhendricks** [comment]
Once we add in the other fixes to our css_parser, we will add it to the builtin plugin/pythonlib library in Sigil and as a first step implement the reformat function in CSSInfo to use the new css_parser to actually fix this issue so we can close it.

We can also reimplement the unused class selector removal code as well.

**Comment by wrCisco** [comment]
I can add the fixes to the problems I stumbled upon these years (the color function case sensitivity matching, some issues with the serializer). About the lack of support for the amazon media types in media queries, in my plugin for css cleaning I just added the appropriate types to the MediaQuery.MEDIA_TYPES list, but I think that the best solution for the parser would be to accept unknown types (since it is clearly foreseen by the spec: https://www.w3.org/TR/css3-mediaqueries/#error-handling). Hoping that the solution won't be too complex, I can work on it tonight and tomorrow.

**Comment by kevinhendricks** [comment]
One approach might be to add a new Sequence to the Choice where you check for NOT  in  MEDIA_TYPES but still meets the type of IDENT.  That said, I would still add the Amazon media types to MEDIA_TYPES just to be safe.

**Comment by wrCisco** [comment]
I just pushed all the updates in a new branch called "serializer".
Changes in summary are:
- two more prefs for the serializer: linesAfterRules (adding blank lines between rules in css) and formatUnknownAtRules (defaults to True for backward compatibility, but letting the parser to format something it doesn't understand is always a risk).
- The serializer should now also comply with the preference indentClosingBrace even with at-rules other than at-media.
- Functions rgb/a and hsl/a now have a case insensitive match.
- MediaQuery now accepts unknown media types and just logs a warning when (and if) sets the mediaType attribute (I follow Kevin's suggestion to add a new Sequence to the productions used for parsing).
- In order to test the warning for the media queries I moved/added some logging facilities from test_errorhandler.py to basetest.py (_setHandler and _captureLog)

If you want, we can add 'amzn-mobi' and 'amzn-kf8' to MEDIA_TYPES, but that can be as easily accomplished at runtime in Sigil or Calibre (at least, I think).

All tests pass both with py2 and py3 and the parser seems to work fine with some real-world ebooks's stylesheets.

**Comment by wrCisco** [comment]
Done, and done.

**Comment by dougmassay** [comment]
> Once we add in the other fixes to our css_parser, we will add it to the builtin plugin/pythonlib library in Sigil and as a first step implement the reformat function in CSSInfo to use the new css_parser to actually fix this issue so we can close it.

The python3lib directory isn't easily accessible to plugins. I propose we make it so for now by adding it to sys.path via launcher.py. I can make sure that any plugin that uses cssutils will automatically import this new module (provide the api is still compatible) via launcher.py as well. This will ensure that anyone using an external Python interpreter for Sigil plugins will still be able to easily take advantage of this new, integral module (think Linux, mostly).

In the future, if we publish the module on PyPI (and I think we probably should), we could drop it from Sigil's source package entirely and make it a third-party requirement (like lxml currently is). Doing so would make it available to our Embedded Python routines as well as Sigil plugins (both in the bundled Python and external Python versions).

**Comment by kevinhendricks** [comment]
I was thinking of treating it like sigil_bs4 for and putting css_parser in the plugin directory where it can be found by EmbeddedPython code if need be.  We can do the hooks if need be to make it look like cssutils in the launcher or simply allow both to co-exist.  And I agree, once css_parser finally reaches PyPI, we can remove it from Sigil source and replace cssutils as a required package.  If we do that should be good to go right?  Or am I missing something?

**Comment by dougmassay** [comment]
You're not missing anything. I am! I'd forgotten all about how we handled the sigil_bs4 module. That should work just fine. We can intercept cssutils and have it import css_parser (or let it coexist as you mentioned) and it will still be available to external plugin interpreters as well. That covers all bases in my opinion.

**Comment by kovidgoyal** [comment]
@wrCisco looks good to me as well. I have opened an uissue on css-parsers issue tracker about publishing on PyPI https://github.com/ebook-utils/css-parser/issues/1

Can I suggest that we move any future discussions about css-parser to its own issue tracker.

**Comment by kevinhendricks** [comment]
We have embeded css_parser inside Sigil and have developed a very preliminary css reformatter in python that is run in Sigil's embedded Python environment.

Initial testing indicates it works.  Much more work will need to be done to replace most of what Misc/CSSInfo.cpp does with equivalents from css_parser

All of this was pushed to master.  So I am finally closing this issue.
Thanks to everyone for all their work, feedback, and help.  Sorry it took so long to get this closed.
